test/pretty-printer/ppTest.ml

```diff
@@ -0,0 +1,69 @@
+open Base
+open Mtt.Ast
+open Mtt.PrettyPrinter
+
+(* QCheck generator for the arbitrary (and most likely invalid) expressions *)
+let generator =
+  QCheck.Gen.(
+    sized
+    @@ fix (fun self size ->
+           let lowercase_id =
+             string_size ~gen:(char_range 'a' 'z') (return 1)
+           in
+           let modal_id id = Mtt.Id.M.mk (id ^ "'") in
+           match size with
+           | 0 ->
+               oneof
+                 [
+                   return Expr.Unit;
+                   (* won't work until expressions with free variables can be pretty-printed *)
+                   (* map
+                      (fun s -> Expr.VarL (Mtt.Id.R.mk s))
+                      (string_size ~gen:(char_range 'a' 'z') (return 1)); *)
+                   map (fun idg -> Expr.VarG (modal_id idg)) lowercase_id;
+                 ]
+           | size ->
+               let open Expr in
+               let unary_node f = map f (self (size - 1)) in
+               let binary_node f = map2 f (self (size / 2)) (self (size / 2)) in
+               oneof
+                 [
+                   binary_node pair;
+                   unary_node fst;
+                   unary_node snd;
+                   binary_node app;
+                   map3 func
+                     (map (fun s -> Mtt.Id.R.mk s) lowercase_id)
+                     (return Type.Unit)
+                     (self (size - 1));
+                   unary_node box;
+                   map3 letbox
+                     (map modal_id lowercase_id)
+                     (self (size / 2))
+                     (self (size / 2));
+                 ]))
+
+(* Primitive expressions printer for testing purposes *)
+let print_ast t = [%sexp_of: Expr.t] t |> Sexp.to_string
+
+let arbitrary_ast = QCheck.make generator ~print:print_ast
```

Comment:
I keep forgetting if I already wrote it but we definitely need a shrinker here. I just tried introducing a bug into the pretty-printer and got this counter-example:
```
Fatal error: exception test `Expression pretty printer preserving syntax` failed on ≥ 1 cases:
(Letbox i'
 (Pair
  (Fst
   (Fun n Unit
    (Snd
     (Pair
      (Letbox b' (Box (App (Snd (VarG x')) (App Unit (VarG o'))))
       (Snd (Snd (App (VarG d') (VarG h')))))
      (Letbox g' (Box (Fun d Unit (Box (VarG o'))))
       (Fun l Unit (Snd (Pair (VarG k') Unit))))))))
  (Box
   (Snd
    (Fun a Unit
     (Snd
      (Letbox b'
       (Snd
        (Pair (App (Letbox c' Unit Unit) (Pair Unit (VarG b')))
         (Pair (Fst Unit) (Fun j Unit (VarG h')))))
       (Pair
        (Fun t Unit
         (Pair (Letbox w' (VarG e') Unit) (Letbox l' (VarG q') (VarG o'))))
        (Fun y Unit (Box (App (VarG s') (VarG c')))))))))))
 (Box
  (Fun l Unit
   (Letbox o'
    (Pair
     (Letbox e'
      (App (Box (App Unit (VarG q')))
       (Letbox s' (App (VarG f') (VarG v')) (Fst (VarG b'))))
      (Pair (Letbox w' (Letbox n' (VarG r') Unit) (Pair (VarG h') (VarG a')))
       (Pair (Fst Unit) (Box Unit))))
     (Fst
      (Letbox y' (Snd (Letbox y' (Fst Unit) (Letbox s' Unit Unit)))
       (Fst (Fst (Fst (VarG e')))))))
    (Fun b Unit
     (Fun o Unit
      (Fst
       (Box
        (App
         (Pair (Letbox s' (Fst (VarG s')) (Fun n Unit Unit))
          (Fun s Unit
           (Pair (Fun p Unit (VarG c')) (Letbox g' (VarG l') Unit))))
         (Snd (Pair (Snd (Pair Unit Unit)) (Snd (Fun e Unit Unit)))))))))))))
```

Reply:
Still this is a cool result we have here!

Reply:
Yeah, I'm implementing it now (so far, it just hangs for some reason)